src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp

```diff
@@ -381,55 +460,84 @@ class ShenandoahHeap : public CollectedHeap, public ShenandoahSpaceInfo {
-  void prepare_regions_and_collection_set(bool concurrent);
   // Evacuation
-  void evacuate_collection_set(bool concurrent);
+  virtual void evacuate_collection_set(bool concurrent);
   // Concurrent root processing
   void prepare_concurrent_roots();
   void finish_concurrent_roots();
   // Concurrent class unloading support
   void do_class_unloading();
   // Reference updating
   void prepare_update_heap_references(bool concurrent);
-  void update_heap_references(bool concurrent);
+  virtual void update_heap_references(bool concurrent);
   // Final update region states
   void update_heap_region_states(bool concurrent);
-  void rebuild_free_set(bool concurrent);
+  virtual void final_update_refs_update_region_states();
 
   void rendezvous_threads(const char* name);
   void recycle_trash();
 public:
+  void rebuild_free_set(bool concurrent);
   void notify_gc_progress();
   void notify_gc_no_progress();
   size_t get_gc_no_progress_count() const;
 
 //
 // Mark support
 private:
-  ShenandoahControlThread*   _control_thread;
+  ShenandoahGeneration*  _global_generation;
+
+protected:
+  ShenandoahController*  _control_thread;
+
+  ShenandoahYoungGeneration* _young_generation;
+  ShenandoahOldGeneration*   _old_generation;
+
+private:
   ShenandoahCollectorPolicy* _shenandoah_policy;
   ShenandoahMode*            _gc_mode;
-  ShenandoahHeuristics*      _heuristics;
   ShenandoahFreeSet*         _free_set;
   ShenandoahPacer*           _pacer;
   ShenandoahVerifier*        _verifier;
 
-  ShenandoahPhaseTimings*    _phase_timings;
-
-  ShenandoahControlThread*   control_thread()          { return _control_thread;    }
+  ShenandoahPhaseTimings*       _phase_timings;
+  ShenandoahMmuTracker          _mmu_tracker;
 
 public:
+  ShenandoahController*   control_thread() { return _control_thread; }
+
+  ShenandoahGeneration*      global_generation() const { return _global_generation; }
+  ShenandoahYoungGeneration* young_generation()  const {
+    assert(mode()->is_generational(), "Young generation requires generational mode");
+    return _young_generation;
+  }
+
+  ShenandoahOldGeneration*   old_generation()    const {
+    assert(mode()->is_generational(), "Old generation requires generational mode");
+    return _old_generation;
+  }
+
+  ShenandoahGeneration*      generation_for(ShenandoahAffiliation affiliation) const;
+
   ShenandoahCollectorPolicy* shenandoah_policy() const { return _shenandoah_policy; }
   ShenandoahMode*            mode()              const { return _gc_mode;           }
-  ShenandoahHeuristics*      heuristics()        const { return _heuristics;        }
   ShenandoahFreeSet*         free_set()          const { return _free_set;          }
   ShenandoahPacer*           pacer()             const { return _pacer;             }
 
-  ShenandoahPhaseTimings*    phase_timings()     const { return _phase_timings;     }
+  ShenandoahPhaseTimings*      phase_timings()   const { return _phase_timings;     }
```

Comment:
I'll fix this, but I secretly don't like this style of formatting because it is slightly tedious to maintain. It also suggests that these methods are somehow grouped and formatted together for a reason beyond aesthetics.

Reply:
https://bugs.openjdk.org/browse/JDK-8341992